### PR TITLE
Overhaul repo to support plugins

### PR DESCRIPTION
**Pull request by collindutter**
Overhauls the repo to support more than just Tools. Would like feedback on:
1. Does the term "plugin" make sense? "Extension"?
2. Do the README instructions make sense? Would you know how to create a Prompt Driver?

### CODE REVIEW

**Comment by collindutter** [comment]
3.12 failing is expected, it's already fixed on `griptape@dev`

File: .github/actions/init-environment/action.yml [hunk 1]
```diff
@@ -20,18 +21,17 @@ runs:
...
         if: steps.cached-poetry-dependencies.outputs.cache-hit != 'true'
-        run: poetry install --no-interaction --no-root --with test --with dev --all-extras
+        run: poetry install --no-interaction --with test --with dev --all-extras
         shell: bash
-
-      - name: Activate venv
+      - name: Activate venv 
         run: |
-          source .venv/bin/activate
+          source $VENV
```
Comment (dylanholmes) [inline .github/actions/init-environment/action.yml:35]: TIL

File: .github/workflows/unit-tests.yml [hunk 2]
```diff
@@ -2,21 +2,25 @@ name: Unit Tests
...
-    branches: [ "main" ]
+    branches: ["main"]
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: true
 
 jobs:
-  test:
+  test-ubuntu:
```
Comment (dylanholmes) [inline .github/workflows/unit-tests.yml:14]: Windows too?

**Review by dylanholmes — COMMENTED** [review]
I'm fine with "plugin".

I think it's pretty straight forward on how you'd add a prompt driver, but you could provide links to griptape maintained plugin repos for more in depth examples (once we have them).

File: README.md [hunk 3]
```diff
@@ -1,17 +1,106 @@
-# Griptape Tool Template
+# Griptape Plugins Template
 
-A Github template repository for creating Griptape tools.
+A Github template repository for creating Griptape plugins.
```
Comment (SavagePencil) [inline README.md:3]: Suggest "extensions" as it implies extending functionality and capabilities. Opinion, tho.

File: README.md [hunk 4]
```diff
@@ -1,17 +1,106 @@
...
-### Steps for creating a new Griptape tool:
+```
+$ gh repo create griptape-plugin-name -p griptape/griptape-plugin-template
+```
+
+Via github web page:
+
+Click on `Use this template`
+
+![](https://docs.github.com/assets/cb-36544/images/help/repository/use-this-template-button.png)
```
Comment (SavagePencil) [inline README.md:17]: Make this one first. Everyone here will know that one, not everyone will be as familiar with the CLI

File: README.md [hunk 5]
```diff
@@ -1,17 +1,106 @@
...
+
+## What is a Griptape Plugin?
+
+A Griptape Plugin is nothing more than a python class that implements one of Griptape's interfaces.
```
Comment (SavagePencil) [inline README.md:21]: Start with what it affords, e.g., "A Griptape [extension/plugin] allows you to make your own functionality for your Griptape agents..." Since you are already using the 2nd person elsewhere in this document, the second sentence can be "You create one as a Python class that implements a Griptape interface."

You've told me what one is but now why I would want to use one. Give some concrete examples.

File: README.md [hunk 6]
```diff
@@ -1,17 +1,106 @@
...
+```bash
+make check
+```
+
+### Running Formatter
+
+```bash
+make format
+```
+
```
Comment (SavagePencil) [inline README.md:81]: Should these be at `####` (four indents) as they appear to be subsets of `make setup`

File: README.md [hunk 7]
```diff
@@ -1,17 +1,106 @@
...
+## Development
+
+### One Command Setup
+
+Run the following command to do a one-time setup of the project:
+
+```bash
+make setup
+```
+
```
Comment (SavagePencil) [inline README.md:55]: Summarize what that results in (e.g., "This will install all dependencies, run tests to validate no regressions, run checks (what is that?), and run the formatter (why isn't it formatted already?).

File: README.md [hunk 8]
```diff
@@ -1,17 +1,106 @@
...
+### Running Formatter
+
+```bash
+make format
+```
+
+### Running Example
+
+1. Set appropriate environment variables. The example requires an `OPENAI_API_KEY` environment variable to be set.
```
Comment (SavagePencil) [inline README.md:84]: can you do a `1a.` that says "each example will require its own environment variables. See the XXX in each example to understand which environment variables are required to run them."

File: README.md [hunk 9]
```diff
@@ -1,17 +1,106 @@
...
+2. Run the example:
+
+```bash
+poetry run python examples/example.py
+```
+
+### Installing In Other Projects
+
```
Comment (SavagePencil) [inline README.md:92]: Talk about why and when you would want to do this, possibly with an example.

**Review by SavagePencil — CHANGES_REQUESTED** [review]
We've said what but not why. Small inserts can fix that.

File: README.md (hunk 4 above)
Comment (shhlife) [reply README.md:17]: agreed

File: README.md (hunk 5 above)
Comment (shhlife) [reply README.md:21]: I understand the reason behind saying "nothing more than" because it makes it sound simple and easy.. but it ends up sounding not very impressive.  Maybe just simplify?

also, not sure what "implements one of Griptape's interfaces" means. When I hear "interfaces" I think "UI".

File: README.md [hunk 10]
```diff
@@ -1,17 +1,106 @@
...
+make format
+```
+
+### Running Example
+
+1. Set appropriate environment variables. The example requires an `OPENAI_API_KEY` environment variable to be set.
+2. Run the example:
+
+```bash
+poetry run python examples/example.py
```
Comment (shhlife) [inline README.md:88]: This assumes the user is using `poetry` - should we make that assumption? do we want to include `pip` instructions as well?

File: README.md [hunk 11]
```diff
@@ -1,17 +1,106 @@
...
+
+A Griptape Plugin is nothing more than a python class that implements one of Griptape's interfaces.
+
+This repository provides a recommended structure for organizing your plugin code, as well as helpful tools for testing and development.
+
+## Plugin Structure
+
+The template repository is structured as follows:
+
+├── griptape
```
Comment (shhlife) [inline README.md:29]: This wasn't showing as a clean hierarchy in the preview for me - do you want to wrap it in ```?

**Review by shhlife — CHANGES_REQUESTED** [review]
made a couple of notes

File: README.md (hunk 3 above)
Comment (collindutter) [reply README.md:3]: I think I like extension more. Plugin, to me, suggests some sort of modification of core behavior. This template is for providing new, extended functionality.

File: README.md (hunk 10 above)
Comment (collindutter) [reply README.md:88]: This repo is heavily dependent on poetry. I think it's fine as long we provide some justification for it. I'll include some more context about what poetry is and why we're using it.

File: README.md [hunk 12]
```diff
@@ -1,17 +1,119 @@
...
+```
+
+#### Running Formatter
+
+```bash
+make format
+```
+
+#### Running Example
+
```
Comment (SavagePencil) [inline README.md:92]: 1. We gave you an example you can run
2. Here's what it does in plain English
3. Here's what role the plugin/extension plays in that experience
4. Here's where to look for when you want to learn how the plugin/extension was built and hooked up

File: README.md [hunk 13]
```diff
@@ -1,17 +1,119 @@
...
+Via `gh`:
+
+```
+$ gh repo create griptape-extension-name -p griptape/griptape-extension-template
+```
+
+## What is a Griptape Extension?
+
+Griptape Extensions can add new functionality to the [griptape framework](https://github.com/griptape-ai/griptape), such as new Tools, Drivers, Tasks, or Structures.
```
Comment (SavagePencil) [inline README.md:22]: Be clear what that affords. "With Griptape Extensions, your Agents can now do X, Y, Z."

Also capitalize `Griptape framework`

File: README.md [hunk 14]
```diff
@@ -1,17 +1,119 @@
...
+1. Set appropriate environment variables. The example requires an `OPENAI_API_KEY` environment variable to be set.
+2. Run the example:
+
+```bash
+poetry run python examples/example.py
+```
+
+### Installing in Other Projects
+
+Extensions are designed to be shared and installed in projects that can benefit from them.
```
Comment (SavagePencil) [inline README.md:102]: How about `Extensions are designed to be shared. Extensions are made to easily install (include? import?) into existing Python projects.`

File: README.md [hunk 15]
```diff
@@ -1,17 +1,119 @@
...
+
+Although publishing to [PyPI](https://pypi.org/) is an option, it is beyond the scope of this template. Instead, extensions can be installed directly from the repository.
```
Comment (SavagePencil) [inline README.md:104]: I'd remove the PyPi comment, because it gives the reader a sense of dread (n.b.: I am always sensing dread). Be optimistic: `The easiest way to include your extension into an existing project is to install directly from the repository, like so:`

...

later:
`Advanced customers may seek to publish their extensions to PyPi. Those instructions are beyond the scope of this README.`

**Review by SavagePencil — CHANGES_REQUESTED** [review]
get ahead...ahead of that dread!

File: README.md (hunk 13 above)
Comment (collindutter) [reply README.md:22]: What do you have in mind for X, Y, Z? Extensions don't do anything concrete on their own, they simply enable the user to develop functionality like new Tools, Drivers, Tasks, or Structures.

Comment (SavagePencil) [reply README.md]: right, but that may not be obvious what that affords. "With extensions, your Griptape Agents can leverage your team's APIs, perform custom business logic, and access proprietary data sources, all as part of their reasoning."

Comment (collindutter) [reply README.md]: Added a sentence.
